AK/LexicalPath.cpp

```diff
@@ -12,85 +12,102 @@
+                    }
+                } else {
+                    if (canonical_parts.last() != "..") {
+                        // A .. and a previous non-.. part cancel each other.
+                        canonical_parts.take_last();
+                        continue;
+                    }
                 }
             }
-        }
-        if (!part.is_empty()) {
             approximate_canonical_length += part.length() + 1;
             canonical_parts.append(part);
         }
-    }
-    if (canonical_parts.is_empty()) {
-        m_string = m_basename = m_dirname = m_is_absolute ? "/" : ".";
-        return;
-    }
 
-    StringBuilder dirname_builder(approximate_canonical_length);
-    for (size_t i = 0; i < canonical_parts.size() - 1; ++i) {
-        auto& canonical_part = canonical_parts[i];
-        if (m_is_absolute || i != 0)
-            dirname_builder.append('/');
-        dirname_builder.append(canonical_part);
-    }
-    m_dirname = dirname_builder.to_string();
+        if (canonical_parts.is_empty() && !is_absolute())
+            canonical_parts.append(".");
 
-    if (m_dirname.is_empty()) {
-        m_dirname = m_is_absolute ? "/" : ".";
+        StringBuilder builder(approximate_canonical_length);
+        if (is_absolute())
+            builder.append('/');
+        builder.join('/', canonical_parts);
+        m_string = builder.to_string();
     }
 
-    m_basename = canonical_parts.last();
+    m_parts = m_string.split_view('/');
 
-    Optional<size_t> last_dot = StringView(m_basename).find_last_of('.');
-    if (last_dot.has_value()) {
-        m_title = m_basename.substring(0, last_dot.value());
-        m_extension = m_basename.substring(last_dot.value() + 1, m_basename.length() - last_dot.value() - 1);
+    auto last_slash_index = m_string.view().find_last_of('/');
+    if (!last_slash_index.has_value()) {
+        // The path contains a single part and is not absolute. m_dirname = "."sv
+        m_dirname = { &s_single_dot, 1 };
+    } else if (*last_slash_index == 0) {
+        // The path contains a single part and is absolute. m_dirname = "/"sv
+        m_dirname = m_string.substring_view(0, 1);
     } else {
-        m_title = m_basename;
+        m_dirname = m_string.substring_view(0, *last_slash_index);
     }
 
-    StringBuilder builder(approximate_canonical_length);
-    for (size_t i = 0; i < canonical_parts.size(); ++i) {
-        auto& canonical_part = canonical_parts[i];
-        if (m_is_absolute || i != 0)
-            builder.append('/');
-        builder.append(canonical_part);
+    if (m_string == "/")
+        m_basename = m_string;
+    else {
+        VERIFY(m_parts.size() > 0);
+        m_basename = m_parts.last();
+    }
+
+    auto last_dot_index = m_basename.find_last_of('.');
+    // NOTE: if the dot index is 0, this means we have ".foo", it's not an extension, as the title would then be "".
+    if (last_dot_index.has_value() && *last_dot_index != 0) {
```

Comment:
```suggestion
    if (last_dot_index != 0u) {
```

Reply:
Wouldn't this evaluate to `true` if if doesn't have a value?

Reply:
Ah fair.
My bad.